### PR TITLE
Atomic host unlock test

### PR DESCRIPTION
**Pull request by chuanchang**
This PR is based on #90, please review #90 firstly, and then I will rebase the PR once #90 is merged.

steps/common.py: run command ignore error
steps/rhelah.py: update mount related functions and add atomic host unlock function
steps/ostree.py: define get_ostree_admin_status()

features/rhelah/atomic/atomic_mount.feature,
features/rhelah/atomic/sanity_test.feature: update mount related testing

features/rhelah/atomic/atomic_host_unlock.feature: add testing for atomic host unlock

``` shell
$ behave features/rhelah/atomic/atomic_host_unlock.feature
Feature: Atomic host unlock sanity test # features/rhelah/atomic/atomic_host_unlock.feature:2
  Describes the basic 'atomic host unlock' command test
  Background: Atomic hosts are discovered  # features/rhelah/atomic/atomic_host_unlock.feature:5

  @change_to_development_mode
  Scenario: 1. change atomic host to development mode               # features/rhelah/atomic/atomic_host_unlock.feature:9
    Given "all" hosts can be pinged                                 # steps/common.py:205 1.666s
    When Switch atomic host to development mode                     # steps/rhelah.py:793 0.917s
    When Create file "/usr/unlock_dev_test"                         # steps/rhelah.py:812 0.253s
    Then Check unlock status "Unlocked: development" exists         # steps/rhelah.py:801 0.162s
    And Check whether mount option "overlay /usr overlay rw" exists # steps/rhelah.py:519 0.144s
    And run command "shell ls /usr/unlock_dev_test"                 # steps/common.py:213 0.134s

  @reboot_host_then_check_dev_mode
  Scenario: 2. reboot atomic host then check development mode               # features/rhelah/atomic/atomic_host_unlock.feature:17
    Given "all" hosts can be pinged                                         # steps/common.py:205 0.142s
    When wait "60" seconds for "all" to reboot                              # steps/rhelah.py:170 62.078s
    Then Check unlock status "Unlocked: development" does not exists "true" # steps/rhelah.py:801 0.338s
    And check whether mount option "overlay /usr overlay rw" does not exist # steps/rhelah.py:536 0.241s
    And run command "shell ls /usr/unlock_test" ignore error "true"         # steps/common.py:213 0.258s

  @change_to_hotfix_mode
  Scenario: 3. change atomic host to hotfix mode                    # features/rhelah/atomic/atomic_host_unlock.feature:24
    Given "all" hosts can be pinged                                 # steps/common.py:205 0.228s
    When Switch atomic host to "hotfix" mode                        # steps/rhelah.py:793 13.135s
    When Create file "/usr/unlock_hotfix_test"                      # steps/rhelah.py:812 0.244s
    Then Check unlock status "Unlocked: hotfix" exists              # steps/rhelah.py:801 0.152s
    And Check whether mount option "overlay /usr overlay rw" exists # steps/rhelah.py:519 0.149s
    And run command "shell ls /usr/unlock_hotfix_test"              # steps/common.py:213 0.145s

  @reboot_host_then_check_hotfix_mode
  Scenario: 4. reboot atomic host then check hotfix mode            # features/rhelah/atomic/atomic_host_unlock.feature:32
    Given "all" hosts can be pinged                                 # steps/common.py:205 0.126s
    When wait "60" seconds for "all" to reboot                      # steps/rhelah.py:170 61.797s
    Then Check unlock status "Unlocked: hotfix" exists              # steps/rhelah.py:801 0.219s
    And check whether mount option "overlay /usr overlay rw" exists # steps/rhelah.py:519 0.149s
    And run command "shell ls /usr/unlock_hotfix_test"              # steps/common.py:213 0.152s

  @rollback_host_after_hotfix_mode
  Scenario: 5. Rollback to the original deployment                           # features/rhelah/atomic/atomic_host_unlock.feature:39
    Given "all" hosts can be pinged                                          # steps/common.py:205 0.135s
    When atomic host rollback is successful                                  # steps/rhelah.py:368 5.421s
    And wait "60" seconds for "all" to reboot                                # steps/rhelah.py:170 62.022s
    Then check whether mount option "overlay /usr overlay rw" does not exist # steps/rhelah.py:536 0.237s
    And run command "shell ls /usr/unlock_test" ignore error "true"          # steps/common.py:213 0.262s

  @rollback_host_to_hotfix_mode
  Scenario: 6. Rollback to the hotfix deployment                    # features/rhelah/atomic/atomic_host_unlock.feature:46
    Given "all" hosts can be pinged                                 # steps/common.py:205 0.261s
    When atomic host rollback is successful                         # steps/rhelah.py:368 5.489s
    And wait "60" seconds for "all" to reboot                       # steps/rhelah.py:170 61.858s
    Then Check unlock status "Unlocked: hotfix" exists              # steps/rhelah.py:801 0.210s
    And Check whether mount option "overlay /usr overlay rw" exists # steps/rhelah.py:519 1.138s
    And run command "shell ls /usr/unlock_hotfix_test"              # steps/common.py:213 0.147s

  @rollback_host
  Scenario: 7. Rollback to the original deployment                           # features/rhelah/atomic/atomic_host_unlock.feature:54
    Given "all" hosts can be pinged                                          # steps/common.py:205 0.135s
    When atomic host rollback is successful                                  # steps/rhelah.py:368 4.925s
    And wait "60" seconds for "all" to reboot                                # steps/rhelah.py:170 62.082s
    Then check whether mount option "overlay /usr overlay rw" does not exist # steps/rhelah.py:536 0.237s
    And run command "shell ls /usr/unlock_test" ignore error "true"          # steps/common.py:213 0.232s

1 feature passed, 0 failed, 0 skipped
7 scenarios passed, 0 failed, 0 skipped
38 steps passed, 0 failed, 0 skipped, 0 undefined
Took 5m47.620s
```

### CODE REVIEW

File: features/rhelah/atomic/atomic_host_unlock.feature [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        Then check whether mount option "overlay /usr overlay rw" does not exist
+         and run command "shell ls /usr/unlock_test" ignore error "true"
+
+  @rollback_host_to_hotfix_mode
+  Scenario: 6. Rollback to the hotfix deployment
+        When atomic host rollback is successful
+         and wait "60" seconds for "all" to reboot
+        Then Check unlock status "Unlocked: hotfix" exists
+         and Check whether mount option "overlay /usr overlay rw" exists
+         and run command "shell ls /usr/unlock_hotfix_test"
```
Comment (miabbott) [inline features/rhelah/atomic/atomic_host_unlock.feature]: nit: the indent of the `When` and `Then` statements between lines 10-51 is off

**Comment by chuanchang** [comment]
Fix above errors later.

**Comment by chuanchang** [comment]
@miabbott, I will rebase this once the PR90 is merged.

**Comment by chuanchang** [comment]
@miabbott, please help review, thanks in advance.

File: features/rhelah/atomic/atomic_host_unlock.feature [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        Then check whether mount option "overlay /usr overlay rw" does not exist
+         and run command "shell ls /usr/unlock_test" ignore error "true"
+
+  @rollback_host_to_hotfix_mode
+  Scenario: 6. Rollback to the hotfix deployment
+        When atomic host rollback is successful
+         and wait "60" seconds for "all" to reboot
+        Then check unlock status "Unlocked: hotfix" exists
+         and Check whether mount option "overlay /usr overlay rw" exists
+         and run command "shell ls /usr/unlock_hotfix_test"
```
Comment (miabbott) [inline features/rhelah/atomic/atomic_host_unlock.feature]: nit: the `When` and `Then` statements in lines 10 through 51 are indented too much

File: steps/ostree.py [hunk 3]
```diff
@@ -4,14 +4,16 @@
...
     ostree_result = context.remote_cmd(cmd='command',
                                        module_args='ostree admin status')
+    return ostree_result[0]['stdout']
 
-    assert ostree_result, "Can not get ostree status"
 
-    ostree_status = ostree_result[0]['stdout']
+@then(u'undeploy the unselected deployment')
+def step_impl(context):
+    ostree_status = get_ostree_admin_status(context)
```
Comment (miabbott) [inline steps/ostree.py:19]: I think you should check that `ostree_status` has valid value before trying to use it.

File: steps/rhelah.py [hunk 4]
```diff
@@ -785,3 +786,29 @@ def step_impl(context, container, mount_option, option=None):
...
+@then(u'check unlock status "{mode_str}" exists')
+@then(u'check unlock status "{mode_str}" does not exists "{ignore_rc}"')
+def step_impl(context, mode_str, ignore_rc='false'):
+    '''Check unlock status'''
+    rc = string_to_bool(context, ignore_rc)
+    ostree_status = get_ostree_admin_status(context)
+    if rc:
+        assert mode_str not in ostree_status, "Can still find '%s' in %s" %(mode_str, ostree_status)
+    else:
+        assert mode_str in ostree_status, "Can't find '%s' in %s" %(mode_str, ostree_status)
```
Comment (miabbott) [inline steps/rhelah.py]: nit: need a space between the `%` and the tuple in the previous two lines

File: steps/rhelah.py [hunk 5]
```diff
@@ -785,3 +786,29 @@ def step_impl(context, container, mount_option, option=None):
...
+    opt_dict = {"development": "", "hotfix": "--hotfix"}
+    assert context.remote_cmd(cmd='command',
+                              module_args='atomic host unlock %s' % opt_dict[mode])
+
+@then(u'check unlock status "{mode_str}" exists')
+@then(u'check unlock status "{mode_str}" does not exists "{ignore_rc}"')
+def step_impl(context, mode_str, ignore_rc='false'):
+    '''Check unlock status'''
+    rc = string_to_bool(context, ignore_rc)
+    ostree_status = get_ostree_admin_status(context)
```
Comment (miabbott) [inline steps/rhelah.py:803]: Same note here about checking for a valid value in `ostree_status` before trying to use it

**Comment by miabbott** [comment]
@chuanchang Good test!  A couple of style fixes and some code suggestions for you to consider.

**Comment by chuanchang** [comment]
@miabbott, done, thanks :)

**Comment by miabbott** [comment]
LGTM!

**Pull request merged**